### PR TITLE
Inner packet hashing test plan

### PR DESCRIPTION
**Pull request by anish-n**

### CODE REVIEW

File: doc/ecmp/inner_packet_hashing_test_plan.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+### Vxlan inner packet hashing
+The test will send vxlan inner packets and validate that varying any single 5-tuple of the inner packet results in hash variation.
+Vxlan packets will be sent with both outer IPv4 and outer IPv6.
+
+### Inner packet types for tuple variation
+Validate hashing with different inner packet types:
+1. IPv4 TCP
+2. IPv4 UDP
+3. IPv4 ICMP
```
Comment (nazariig) [inline doc/ecmp/inner_packet_hashing_test_plan.md:36]: @anish-n What about inner frames with IPv6 (when `IP protocol` = `Next Header`)?

File: doc/ecmp/inner_packet_hashing_test_plan.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+### High level test details
+1. Send packets to a destination prefix which is pointing to multiple ecmp nexthops. For the T0 topology test we will send it to a dest prefix which is pointing to the T1s.
+3. Vary some tuple of the packet so that the packets hash to different nexthops. The total packets sent in a test is calcualted as follows: 1000 packets sent per ECMP next hop. This translates to 4000 packets in a T0 topology with 4 T1s(ECMP nexthops). All 4000 packets will have varied tuples to get a good distribution of packets to ports.
+4. Identify set of ports on which the packet would have ecmp'd
+5. Check which port received the packet and record received packet count per port
+6. Calculate the expected number of packets per port
+7. Validate if received packet count per port is within a 25% deviation of expected number of packets on the port
+
+### Vxlan inner packet hashing
```
Comment (nazariig) [inline doc/ecmp/inner_packet_hashing_test_plan.md:28]: @anish-n What about NVGRE?

File: doc/ecmp/inner_packet_hashing_test_plan.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+Currently this test expects that the inner packet hashing is pre-configured on the DUT. In the future, once dynamic Policy Based Hashing feature is implemented, dynamic configuration can be done on the DUT and packet hashing verified per the config.
+
+Beyond hash configuration, base DUT route table configurations are sufficient to perform the test.
+
+### Supported topology
+The test will be supported on the T0 toplogy, it may be enhanced in the future for other topologies.
+
+### High level test details
+1. Send packets to a destination prefix which is pointing to multiple ecmp nexthops. For the T0 topology test we will send it to a dest prefix which is pointing to the T1s.
+3. Vary some tuple of the packet so that the packets hash to different nexthops. The total packets sent in a test is calcualted as follows: 1000 packets sent per ECMP next hop. This translates to 4000 packets in a T0 topology with 4 T1s(ECMP nexthops). All 4000 packets will have varied tuples to get a good distribution of packets to ports.
```
Comment (nazariig) [inline doc/ecmp/inner_packet_hashing_test_plan.md:22]: @anish-n T0 has 4 LAGs with a single member per LAG and VLAN with 24 members. Does it mean that we are going to send 1000 packets per LAG and expect packets will be distributed across VLAN members?

File: doc/ecmp/inner_packet_hashing_test_plan.md (hunk 1 above)
Comment (anish-n) [reply doc/ecmp/inner_packet_hashing_test_plan.md:36]: Inner IPv6 is out of scope, hence included only inner iPv4 types

File: doc/ecmp/inner_packet_hashing_test_plan.md (hunk 2 above)
Comment (anish-n) [reply doc/ecmp/inner_packet_hashing_test_plan.md:28]: Good point, will include NVGRE as possible outer encap format here

File: doc/ecmp/inner_packet_hashing_test_plan.md (hunk 3 above)
Comment (anish-n) [reply doc/ecmp/inner_packet_hashing_test_plan.md:22]: No the direction of packets is the opposite, ~4000 packets sent from the VLAN members and received at the T1, we will check the distribution at the T1

File: doc/ecmp/inner_packet_hashing_test_plan.md (hunk 1 above)
Comment (nazariig) [reply doc/ecmp/inner_packet_hashing_test_plan.md]: @anish-n please specify this explicitly: inner frames are IPv4 only

File: doc/ecmp/inner_packet_hashing_test_plan.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+### Different outer encapsulation formats tested with various inner packets
+1. The test will send inner packets and validate that varying any single 5-tuple of the inner packet results in hash variation, it will be sent with various outer encapsulation formats listed below
+2. IPv4 Vxlan
+3. IPv6 Vxlan
+4. IPv4 NVGRE
+5. IPv6 NVGRE
+
+### Inner packet tuples varied for hash valdiation:
+Validate hashing with different inner packet tuples:
```
Comment (nazariig) [inline doc/ecmp/inner_packet_hashing_test_plan.md:36]: @anish-n please specify explicitly: inner frames are IPv4 only

File: doc/ecmp/inner_packet_hashing_test_plan.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
+# Test Plan for inner packet hashing in ECMP
+###### Revision
+| Rev |     Date    |       Author       | Change Description                |
+|:---:|:-----------:|:------------------:|-----------------------------------|
+| 0.1 | 03/12/2021  |    Anish Narsian   | Initial version                   |
+
+## Overview
+The purpose of this test plan is to describe inner packet hashing tests for ECMP nexthops. The test methodology will be similar to the current hash_test in SONiC. The test sends vxlan encapsulated packets and validates hashing is performed per the inner packet's 5 tuple(src ip, dst ip, src port, dst port, ip proto).
```
Comment (nazariig) [inline doc/ecmp/inner_packet_hashing_test_plan.md:8]: @anish-n Suggest to change `The test sends vxlan encapsulated packets` -> `The test sends encapsulated packets`

**Comment by nazariig** [comment]
@anish-n and what about warm/fast-reboot testing?

File: doc/ecmp/inner_packet_hashing_test_plan.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+7. Validate if received packet count per port is within a 25% deviation of expected number of packets on the port
+
+### Different outer encapsulation formats tested with various inner packets
+1. The test will send inner packets and validate that varying any single 5-tuple of the inner packet results in hash variation, it will be sent with various outer encapsulation formats listed below
+2. IPv4 Vxlan
+3. IPv6 Vxlan
+4. IPv4 NVGRE
+5. IPv6 NVGRE
+
+### Inner packet tuples varied for hash valdiation:
```
Comment (nazariig) [inline doc/ecmp/inner_packet_hashing_test_plan.md:35]: @anish-n do we need to verify that hashing works only for inner 5-tuple and not for other parameters?

Comment (anish-n) [reply doc/ecmp/inner_packet_hashing_test_plan.md:35]: You mean that outer packet params should not inadvertently contribute to the hashing? So we need to make sure that all of the outer packet params are kept constant but the packet still gets varied across nhs given inner packet variance.. That was an implicit case, but its good to make it explicit, thanks, will add that.

File: doc/ecmp/inner_packet_hashing_test_plan.md (hunk 5 above)
Comment (anish-n) [reply doc/ecmp/inner_packet_hashing_test_plan.md]: Done, thanks

File: doc/ecmp/inner_packet_hashing_test_plan.md (hunk 4 above)
Comment (anish-n) [reply doc/ecmp/inner_packet_hashing_test_plan.md]: Updated

**Comment by anish-n** [comment]
> @anish-n and what about warm/fast-reboot testing?

Added, thanks

**Review by nazariig — APPROVED** [review]

**Review by prsunny — APPROVED** [review]

**Pull request merged**